src/gui/qgsmaptoolselectionhandler.cpp

```diff
@@ -0,0 +1,460 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmaptoolselectionhandler.h"
+
+#include <QBoxLayout>
+#include <QKeyEvent>
+#include <QLabel>
+
+#include "qgisinterface.h"
+#include "qgsdoublespinbox.h"
+#include "qgsmapcanvas.h"
+#include "qgsmapmouseevent.h"
+#include "qgsrubberband.h"
+#include "qgssnapindicator.h"
+
+
+/// @cond private
+
+QgsDistanceWidget::QgsDistanceWidget( const QString &label, QWidget *parent )
+  : QWidget( parent )
+{
+  mLayout = new QHBoxLayout( this );
+  mLayout->setContentsMargins( 0, 0, 0, 0 );
+  mLayout->setAlignment( Qt::AlignLeft );
+  setLayout( mLayout );
+
+  if ( !label.isEmpty() )
+  {
+    QLabel *lbl = new QLabel( label, this );
+    lbl->setAlignment( Qt::AlignRight | Qt::AlignCenter );
+    mLayout->addWidget( lbl );
+  }
+
+  mDistanceSpinBox = new QgsDoubleSpinBox( this );
+  mDistanceSpinBox->setSingleStep( 1 );
+  mDistanceSpinBox->setValue( 0 );
+  mDistanceSpinBox->setMinimum( 0 );
+  mDistanceSpinBox->setMaximum( 1000000000 );
+  mDistanceSpinBox->setDecimals( 6 );
+  mDistanceSpinBox->setShowClearButton( false );
+  mDistanceSpinBox->setSizePolicy( QSizePolicy::MinimumExpanding, QSizePolicy::Preferred );
+  mLayout->addWidget( mDistanceSpinBox );
+
+  // connect signals
+  mDistanceSpinBox->installEventFilter( this );
+  connect( mDistanceSpinBox, static_cast < void ( QgsDoubleSpinBox::* )( double ) > ( &QgsDoubleSpinBox::valueChanged ), this, &QgsDistanceWidget::distanceChanged );
+
+  // config focus
+  setFocusProxy( mDistanceSpinBox );
+}
+
+void QgsDistanceWidget::setDistance( double distance )
+{
+  mDistanceSpinBox->setValue( distance );
+  mDistanceSpinBox->selectAll();
+}
+
+double QgsDistanceWidget::distance()
+{
+  return mDistanceSpinBox->value();
+}
+
+bool QgsDistanceWidget::eventFilter( QObject *obj, QEvent *ev )
+{
+  if ( obj == mDistanceSpinBox && ev->type() == QEvent::KeyPress )
+  {
+    QKeyEvent *event = static_cast<QKeyEvent *>( ev );
+    if ( event->key() == Qt::Key_Escape )
+    {
+      emit distanceEditingCanceled();
+      return true;
+    }
+    if ( event->key() == Qt::Key_Enter || event->key() == Qt::Key_Return )
+    {
+      emit distanceEditingFinished( distance(), event->modifiers() );
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/// @endcond
+
+
+
+QgsMapToolSelectionHandler::QgsMapToolSelectionHandler( QgsMapCanvas *canvas, QgsMapToolSelectionHandler::SelectionMode selectionMode )
+  : QObject()
+  , mCanvas( canvas )
+  , mSelectionMode( selectionMode )
+  , mSnapIndicator( qgis::make_unique< QgsSnapIndicator >( canvas ) )
+  , mFillColor( QColor( 254, 178, 76, 63 ) )
+  , mStrokeColor( QColor( 254, 58, 29, 100 ) )
+{
+}
+
+QgsMapToolSelectionHandler::~QgsMapToolSelectionHandler()
+{
+  cancel();
+}
+
+void QgsMapToolSelectionHandler::setInterface( QgisInterface *iface )
+{
+  mQgisInterface = iface;
```

Comment:
What's your opinion on QgisInterface usage in gui classes?
I would prefer a solution with either this class moved to app or with iface interaction delegated to calling code.

Reply:
Actually only now I have realized that after refactoring this class is not needed in gui anymore - let me move it app then...